### PR TITLE
feat: Time based notification

### PR DESCRIPTION
**Pull request by LinMoskovitch**
Implement time based notifications

## Type of change:
- [ ] Bug fix
- [X] New feature
- [X] Fix/new infrastructure

## Background context of the changes:
Main use case in the application.

## The ticket this PR closes:
https://ofirlindekel.atlassian.net/browse/NOTIFY-145?atlOrigin=eyJpIjoiNWNiYjU1NmVlNjc1NDJkN2I4MTA2ZmM2YWNmNzEwMDUiLCJwIjoiaiJ9

## Scenarios that were tested during this implementation:
* Multiple notifications
* Multiple notifications at the same time
* Elapsed time notifications in case of missing (e.g server down)
* Getting notifications when app is closed.

### CODE REVIEW

File: Notify/Notify/Notify/AppShell.xaml.cs [hunk 1]
```diff
@@ -154,22 +157,28 @@ private void setMessagingCenterLocationSubscription()
             });
         }
 
-        private static void updateStatusOfSentNotifications(List<Notification> arrivedLocationNotifications)
+        private static void updateStatusOfSentNotifications(params List<Notification>[] sentNotificationsArrays)
```
Comment (OfirMatasas) [inline Notify/Notify/Notify/AppShell.xaml.cs:160]: Very nice use of params here.

File: Notify/Notify/Notify/AppShell.xaml.cs [hunk 2]
```diff
@@ -154,22 +157,28 @@ private void setMessagingCenterLocationSubscription()
...
         {
             string json = Preferences.Get(Constants.PREFERENCES_NOTIFICATIONS, string.Empty);
             List<Notification> notifications = JsonConvert.DeserializeObject<List<Notification>>(json);
-            
+
             notifications.ForEach(notification =>
             {
-                if (arrivedLocationNotifications.Any(arrivedNotification => arrivedNotification.ID.Equals(notification.ID)))
+                foreach (List<Notification> sentNotifications in sentNotificationsArrays)
```
Comment (OfirMatasas) [inline Notify/Notify/Notify/AppShell.xaml.cs:167]: sentNotificationsList and sentNotificationLists would be a better naming in this issue, don't you think?

File: Notify/Notify/Notify/AppShell.xaml.cs [hunk 3]
```diff
@@ -154,22 +157,28 @@ private void setMessagingCenterLocationSubscription()
...
+                        notification.Status = "Sent";
+                        r_Logger.LogDebug($"Updated status of notification {notification.ID} to 'Sent'");
+                    }
                 }
             });
-            
+
             Preferences.Set(Constants.PREFERENCES_NOTIFICATIONS, JsonConvert.SerializeObject(notifications));
-            AzureHttpClient.Instance.UpdateNotificationsStatus(arrivedLocationNotifications, "Sent");
+            foreach (var sentNotifications in sentNotificationsArrays)
```
Comment (OfirMatasas) [inline Notify/Notify/Notify/AppShell.xaml.cs:178]: Use explicit variable type instead of "var"

File: Notify/Notify/Notify/AppShell.xaml.cs [hunk 4]
```diff
@@ -192,6 +201,41 @@ private List<string> getAllArrivedDestinations(Location location)
...
+            DateTime notificationTime;
+            
+            r_Logger.LogInformation("Getting all elapsed time notifications");
+            
+            elapsedTimeNotifications = notifications
+                .FindAll(
+                    notification =>
+                    {
+                        bool isTimeNotification = notification.Type is NotificationType.Time;
+                        bool isTimeElapsed = false;
```
Comment (OfirMatasas) [inline Notify/Notify/Notify/AppShell.xaml.cs:219]: A space-row should be here, between the variables and the if-statement

File: Notify/Notify/Notify/AppShell.xaml.cs [hunk 5]
```diff
@@ -192,6 +201,41 @@ private List<string> getAllArrivedDestinations(Location location)
...
+                        if (notification.TypeInfo is DateTime time)
+                        {
+                            notificationTime = time;
```
Comment (OfirMatasas) [inline Notify/Notify/Notify/AppShell.xaml.cs:222]: Do we need "notificationTime" variable?
We can only use the "time" variable from the if-statement for the boolean value right afterwards.
Plus, use a different name for the time of the notification than "time" (notificationTime would be good enough :) )

**Review by OfirMatasas — CHANGES_REQUESTED** [review]
Overall very nice work!
Check my comments.

File: Notify/Notify/Notify/AppShell.xaml.cs (hunk 5 above)
Comment (LinMoskovitch) [reply Notify/Notify/Notify/AppShell.xaml.cs:222]: Right, we don't need it. removing it.

**Review by OfirMatasas — APPROVED** [review]
LGTM!

**Pull request merged**